pkg/config/reload.go

```diff
@@ -148,6 +150,16 @@ func (c *Config) ReloadPauseImage(newConfig *Config) error {
 // ReloadPinnedImages updates the PinnedImages with the provided `newConfig`.
 // The method print log in case of any updates.
 func (c *Config) ReloadPinnedImages(newConfig *Config) {
+	if len(newConfig.PinnedImages) == 0 {
+		c.PinnedImages = newConfig.PinnedImages
+		return
+	}
+	sort.Strings(newConfig.PinnedImages)
+	sort.Strings(c.PinnedImages)
+
+	if cmp.Equal(newConfig.PinnedImages, c.PinnedImages) {
+		return
+	}
```

Comment:
This is a nice optimisation here. Correct? This can also be streamlined as follows:

```suggestion
	if cmp.Equal(c.PinnedImages, newConfig.PinnedImages,
		cmpopts.SortSlices(func(a, b string) bool {
			return a < b
		}),
	) {
		return
	}
```

Perhaps it's a bit awkward-looking when wrapped with an if-statement.

Assume that there is "github.com/google/go-cmp/cmp/cmpopts" an import added.

Reply:
That said, this `cmp.Equal()` could probably be used to retire the rest of this function body, correct? I am not sure what the rest of it aims to do, really. It also appears broken, for example:

- Initiate state of pinned_images set using configuration file

```toml
  [crio.image]
    default_transport = "docker://"
    global_auth_file = ""
    pause_image = "registry.k8s.io/pause:3.9"
    pause_image_auth_file = ""
    pause_command = "/pause"
    pinned_images = ["abc", "def"]
    signature_policy = ""
    signature_policy_dir = "/etc/crio/policies"
    image_volumes = "mkdir"
    big_files_temporary_dir = ""
    auto_reload_registries = false
```

- Updated the configuration file and send SIGHUP to reload configuration

```toml
  [crio.image]
    default_transport = "docker://"
    global_auth_file = ""
    pause_image = "registry.k8s.io/pause:3.9"
    pause_image_auth_file = ""
    pause_command = "/pause"
    pinned_images = ["", "", "xyz"]
    signature_policy = ""
    signature_policy_dir = "/etc/crio/policies"
    insecure_registries = ["127.0.0.0/8"]
    image_volumes = "mkdir"
    big_files_temporary_dir = ""
    auto_reload_registries = false
```

What? That can't be right...

There is also a matter of silently injecting pause_image, which makes sense, but should probably be handled a little bit better, for example:

- Initial state, CRI-O was just started

```toml
[crio.image]
default_transport = "docker://"
global_auth_file = ""
pause_image = "registry.k8s.io/pause:3.9"
pause_image_auth_file = ""
pause_command = "/pause"
signature_policy = ""
signature_policy_dir = "/etc/crio/policies"
image_volumes = "mkdir"
big_files_temporary_dir = ""
auto_reload_registries = false
``` 

Note: the pinned_images list is entirely missing.

- After sending SIGHUP to reload current configuration, but there were no new changes provided...

```toml
[crio.image]
default_transport = "docker://"
global_auth_file = ""
pause_image = "registry.k8s.io/pause:3.9"
pause_image_auth_file = ""
pause_command = "/pause"
pinned_images = ["registry.k8s.io/pause:3.9"]
signature_policy = ""
signature_policy_dir = "/etc/crio/policies"
insecure_registries = ["127.0.0.0/8"]
image_volumes = "mkdir"
big_files_temporary_dir = ""
auto_reload_registries = false
```

Hmm, something's afoot here...

We should fix this, I think.

Reply:
That's a good catch 👍 

By the way, I'm thinking about [`regexForPinnedImages`](https://github.com/cri-o/cri-o/blob/main/internal/storage/image.go#L96) and the `pinned_images` list. They might be merged into one list and passed to `imageService`. 

We could build a `pinned_images` list in the `config` package and move the logic below to the right place. Hence `internal/storage/image` would work with a compiled list of `pinned_images` + `pause_image`. 

https://github.com/cri-o/cri-o/blob/747c152817970360fafa2b7264818681436c0939/internal/storage/image.go#L858-L860

@cri-o/cri-o-maintainers WDYT?

Reply:
Something that @LenkaSeg might want to keep a note of, given the other work on configuration reload.